Source/bmalloc/bmalloc/MARRegistry.h

```diff
@@ -0,0 +1,151 @@
+// and full if head ^ tail == size
+const unsigned MARAllocationRecordTableFIFOModulus = 2 * MARTrackedAllocations;
+
+const unsigned PageSize = 16384;
+const unsigned PageShift = 14;
+
+struct BacktraceRecord {
+    unsigned m_numFrames;
+    uint32_t m_hash;
+    void* m_backtraceBuffer[BacktraceMaxSize];
+};
+
+struct MemoryActionRecord {
+    void* address;
+    size_t m_allocationSize;
+    unsigned m_backtraceRegistryIndex;
+    uint32_t m_backtraceHash;
+    bool isAllocation;
+};
+
+typedef struct MARRegistry MARRegistry;
+struct MARRegistry {
+    struct BacktraceRecord m_backtraceRegistry[MARTrackedBacktraces];
+    struct MemoryActionRecord m_allocationRecordTable[MARTrackedAllocations];
+    // push to the tail of the FIFO, evict from head
+    unsigned m_allocationRecordTableHead;
+    unsigned m_allocationRecordTableTail;
+};
+
+struct ExportedAllocationRecord {
+    MARBacktrace m_allocationTrace;
+    MARBacktrace m_deallocationTrace;
+    unsigned m_allocationSize;
+    bool m_isValid;
+};
+
+void MAR_didAllocate(MARRegistry* registry, void* address, size_t size);
+void MAR_didDeallocate(MARRegistry* registry, void* address);
+
+void MAR_recordAllocation(MARRegistry* registry, void* address, size_t size, unsigned numStackFrames, void** backtrace);
+void MAR_recordDeallocation(MARRegistry* registry, void* address, unsigned numStackFrames, void** backtrace);
+struct ExportedAllocationRecord MAR_getAllocationRecord(MARRegistry* registry, void* address);
+
+// Paging helpers
+
+inline void* canonicalizeAddress(void* address)
+{
+    return (void*)((uintptr_t)(address) & ((1ull << 48) - 1));
+}
+
+inline uintptr_t addressToVirtualPageNumber(void* address)
+{
+    return (uintptr_t)(canonicalizeAddress(address)) >> PageShift;
+}
+
+// FIFO helpers
+
+inline unsigned MAR_allocationTableHeadIndex(MARRegistry* registry)
+{
+    return registry->m_allocationRecordTableHead % MARAllocationRecordTableFIFOModulus;
+}
+
+inline unsigned MAR_allocationTableTailIndex(MARRegistry* registry)
+{
+    return registry->m_allocationRecordTableTail % MARAllocationRecordTableFIFOModulus;
+}
+
+inline bool MAR_isAllocationTableFull(MARRegistry* registry)
+{
+    return (registry->m_allocationRecordTableHead ^ registry->m_allocationRecordTableTail) == MARTrackedAllocations;
+}
+
+inline void MAR_incrementAllocationRecordTableHead(MARRegistry* registry)
+{
+    registry->m_allocationRecordTableHead = registry->m_allocationRecordTableHead + 1;
+    registry->m_allocationRecordTableHead %= MARAllocationRecordTableFIFOModulus;
+}
+
+inline void MAR_incrementAllocationRecordTableTail(MARRegistry* registry)
+{
+    registry->m_allocationRecordTableTail = registry->m_allocationRecordTableTail + 1;
+    registry->m_allocationRecordTableHead %= MARAllocationRecordTableFIFOModulus;
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// If the page number % MARProbability == marQualifyingPageIndex, then it qualifies
```

Comment:
Can this be code instead of a comment? a la `bool addressQualifiesForMAR(uintptr_t addr) { return pageno(addr) & MARProbability == marQualifyingPageIndex; }`